pkg/controller/endpoint/endpoints_controller.go

```diff
@@ -659,23 +659,51 @@ func endpointPortFromServicePort(servicePort *v1.ServicePort, portNum int) *v1.E
 	return epp
 }
 
-// overCapacity returns true if there are more addresses in the provided subsets
-// than the maxCapacity.
-func overCapacity(subsets []v1.EndpointSubset) bool {
+// capacityAnnotationSetCorrectly returns false if number of endpoints is greater than maxCapacity or
+// returns true if underCapacity and the annotation is not set.
+func capacityAnnotationSetCorrectly(annotations map[string]string, subsets []v1.EndpointSubset) bool {
 	numEndpoints := 0
 	for _, subset := range subsets {
 		numEndpoints += len(subset.Addresses) + len(subset.NotReadyAddresses)
 	}
-	return numEndpoints > maxCapacity
+	if numEndpoints > maxCapacity {
+		// If subsets are over capacity, they must be truncated so consider
+		// the annotation as not set correctly
+		return false
+	}
+	_, ok := annotations[v1.EndpointsOverCapacity]
+	return !ok
 }
 
-// capacityAnnotationSetCorrectly returns true if overCapacity() is true and the
-// EndpointsOverCapacity annotation is set to "warning" or if overCapacity()
-// is false and the annotation is not set.
-func capacityAnnotationSetCorrectly(annotations map[string]string, subsets []v1.EndpointSubset) bool {
-	val, ok := annotations[v1.EndpointsOverCapacity]
-	if overCapacity(subsets) {
-		return ok && val == "warning"
+// truncateEndpoints ensures that the number of Addresses does not exceed the maxCapacity. If greater
+// than maxCapacity, the addresses will be truncated so that the number of addresses is maxCapacity.
+func truncateEndpoints(endpoints *v1.Endpoints) bool {
+	canBeAdded := maxCapacity
+
+	for i := range endpoints.Subsets {
+		subset := endpoints.Subsets[i]
+		toBeAdded := len(subset.Addresses) + len(subset.NotReadyAddresses)
+		if canBeAdded-toBeAdded >= 0 {
+			canBeAdded -= toBeAdded
+			continue
+		}
+
+		subset.Addresses = truncateAddresses(subset.Addresses, canBeAdded)
+		canBeAdded -= len(subset.Addresses)
+		subset.NotReadyAddresses = truncateAddresses(subset.NotReadyAddresses, canBeAdded)
+
+		endpoints.Subsets[i] = subset
+		endpoints.Subsets = endpoints.Subsets[0 : i+1]
+		return true
 	}
-	return !ok
+	return false
+}
+
+// truncateAddresses takes a list of addresses and returns a truncated list if the length is greater
```

Comment:
Tiny nit: The name `truncateAddresses` implies to me that the addresses will be modified in place (similar to the `truncateEndpoints` func above). With that said, I'm not sure what the appropriate name is, maybe `addressSubset`?

Reply:
I like `addressSubset`. Thanks for the suggestion!